### PR TITLE
Add doc on how to use code paths for custom library

### PR DESCRIPTION
**Pull request by TomeHirata**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/TomeHirata/mlflow/pull/13702?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13702/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13702
```

</p>
</details>

### Related Issues/PRs
N/A

### What changes are proposed in this pull request?

- Add documentation on how to use code paths for custom library

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: docs/source/model/dependencies.rst [hunk 1]
```diff
@@ -326,6 +326,44 @@ Then MLflow will save ``utils.py`` under ``code/`` directory in the model direct
...
+            # .. your prediction logic
+            return prediction
+
+
+    # Log the model
+    with mlflow.start_run() as run:
+        mlflow.pyfunc.log_model(
+            python_model=MyModel(),
+            artifact_path="model",
+            extra_pip_requirements=["code/custom_package.whl"],
```
Comment (WeichenXu123) [inline docs/source/model/dependencies.rst:362]: I think this is not the recommended way.

I recommend to upload the private package wheel file to an internal server or clould storage with a downloading link,
then when logging model, log extra_pip_requirements=[<wheel_link>]

Comment (BenWilson2) [reply docs/source/model/dependencies.rst:362]: Within the note block showing an example object store path for 'once you're ready for production' is ideal, while the example you have here is perfect for dev testing while building a solution.

Comment (TomeHirata) [reply docs/source/model/dependencies.rst:362]: Thank you for the comments both, I'll add the disclaimer.

Comment (TomeHirata) [reply docs/source/model/dependencies.rst:368]: @WeichenXu123 @BenWilson2 I've added the note section, can you take another look?

Comment (TomeHirata) [reply docs/source/model/dependencies.rst:368]: image

File: docs/source/model/dependencies.rst [hunk 2]
```diff
@@ -326,6 +326,50 @@ Then MLflow will save ``utils.py`` under ``code/`` directory in the model direct
...
+            input_example=input_data,
+            code_paths=["custom_package.whl"],
```
Comment (harupy) [inline docs/source/model/dependencies.rst:370]: Does the `code_path` accept wheels? I thought it only accepts `.py` files or directories.

Comment (harupy) [reply docs/source/model/dependencies.rst:370]: cc @WeichenXu123 to confirm.

Comment (TomeHirata) [reply docs/source/model/dependencies.rst:370]: I conducted an experiment on my local with the following code and it works ok, so i guess this usecase is legit.

```python
import mlflow
from mlflow.tracking import MlflowClient
import numpy as np

# from wheel
import sympy

class MyModel(mlflow.pyfunc.PythonModel):
    def predict(self, context, model_input):
        x = [float(sympy.acos(i)) for i in model_input]
        return x

client = MlflowClient()
mlflow.set_tracking_uri("http://127.0.0.1:5000")
experiment_name = "custom_library"
experiment = client.get_experiment_by_name(experiment_name)
if experiment is None:
    experiment_id = client.create_experiment(experiment_name)
else:
    experiment_id = experiment.experiment_id
with mlflow.start_run(experiment_id=experiment_id) as run:
    model_info = mlflow.pyfunc.log_model(
        python_model=MyModel(),
        artifact_path="model",
        input_example=np.array([1.0, 0.0]),
        pip_requirements=["code/sympy-1.13.3-py3-none-any.whl"],
        code_paths=["sympy-1.13.3-py3-none-any.whl"],
    )
loaded_model = mlflow.pyfunc.load_model(model_uri=model_info.model_uri)
print(loaded_model.predict(np.array([1.0, 0.0])))
```

Comment (harupy) [reply docs/source/model/dependencies.rst:370]: Got it. Let's update the `code_paths` docstring. It doesn't sound like it accepts arbitrary files. If we're missing tests, we should also add tests to ensure the example in this PR works.

File: docs/source/model/dependencies.rst [hunk 3]
```diff
@@ -326,6 +326,50 @@ Then MLflow will save ``utils.py`` under ``code/`` directory in the model direct
 When MLflow loads the model for serving, the ``code`` directory will be added to the system path so that you can use the module in your model
 code like ``from utils import my_func``. You can also specify a directory path as ``code_paths`` to save multiple files under the directory:
 
+Use of ``code_paths`` Option for a Custom Library
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To include custom libraries or libraries from a private mirror server when logging your model, the ``code_paths`` option can be used.
```
Comment (harupy) [inline docs/source/model/dependencies.rst:332]:
```suggestion
To include a custom library that is not publicly available on PyPI when logging your model, the ``code_paths`` argument can be used. 
```

File: docs/source/model/dependencies.rst [hunk 4]
```diff
@@ -326,6 +326,50 @@ Then MLflow will save ``utils.py`` under ``code/`` directory in the model direct
 When MLflow loads the model for serving, the ``code`` directory will be added to the system path so that you can use the module in your model
 code like ``from utils import my_func``. You can also specify a directory path as ``code_paths`` to save multiple files under the directory:
 
+Use of ``code_paths`` Option for a Custom Library
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To include custom libraries or libraries from a private mirror server when logging your model, the ``code_paths`` option can be used. 
+This option allows you to upload .whl files or other dependencies alongside your model, ensuring all required libraries are available during serving.
+
+.. note::
```
Comment (harupy) [inline docs/source/model/dependencies.rst:335]:
```suggestion
.. warning::
```

File: docs/source/model/dependencies.rst (hunk 2 above)
Comment (TomeHirata) [reply docs/source/model/dependencies.rst:370]: Sure, let me work on the doc update and potential unit tests in a separate PR!

File: docs/source/model/dependencies.rst [hunk 5]
```diff
@@ -326,6 +326,50 @@ Then MLflow will save ``utils.py`` under ``code/`` directory in the model direct
...
+
+    The following example demonstrates a quick method for including custom libraries for development purposes. 
+    This approach is not recommended for production environments. 
+    For production usage, upload libraries to a custom PyPI server or a cloud storage to ensure reliable and secure access.
+
+::
+
+    my_project/
+    |── train.py
+    └── custom_package.whl
```
Comment (harupy) [inline docs/source/model/dependencies.rst:345]:
```suggestion

<what this is>

::

    my_project/
    |── train.py
    └── custom_package.whl
```

Can we explain what this is?

File: docs/source/model/dependencies.rst [hunk 6]
```diff
@@ -326,6 +326,50 @@ Then MLflow will save ``utils.py`` under ``code/`` directory in the model direct
 When MLflow loads the model for serving, the ``code`` directory will be added to the system path so that you can use the module in your model
 code like ``from utils import my_func``. You can also specify a directory path as ``code_paths`` to save multiple files under the directory:
 
+Use of ``code_paths`` Option for a Custom Library
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To include custom libraries that are not publicly available on PyPI when logging your model, the ``code_paths`` argument can be used. 
+This option allows you to upload .whl files or other dependencies alongside your model, ensuring all required libraries are available during serving.
```
Comment (harupy) [inline docs/source/model/dependencies.rst:333]:
```suggestion
This option allows you to upload ``.whl`` files or other dependencies alongside your model, ensuring all required libraries are available during serving.
```

**Review by harupy — APPROVED** [review]
Left a few more comments, otherwise LGTM!

**Pull request merged**